dali/kernels/signal/wavelet/wavelet_gpu.cu

```diff
@@ -0,0 +1,158 @@
+  // wavelet sample id
+  const int64_t t_id = blockDim.x * blockDim.y * blockIdx.x + b_id;
+  auto& sample = sample_data[blockIdx.z];
+  if (t_id >= sample.size_in) return;
+  __shared__ T shm[1025];
+  auto a = sample.a[blockIdx.y];
+  auto x = std::pow(2.0, a);
+  if (a == 0.0) {
+    shm[b_id] = sample.in[t_id];
+  }
+  else {
+    shm[b_id] = x * sample.in[t_id];
+    shm[1024] = std::pow(2.0, a / 2.0);
+  }
+  for (int i = 0; i < sample.size_b; ++i) {
+    const int64_t out_id = blockIdx.y * sample.size_b * sample.size_in + i * sample.size_in + t_id;
+    auto b = sample.b[i];
+    if (b == 0.0) {
+      sample.out[out_id] = wavelet(shm[b_id]);
+    }
+    else {
+      sample.out[out_id] = wavelet(shm[b_id] - b);
+    }
+    if (a != 0.0) {
+      sample.out[out_id] *= shm[1024];
+    }
+  }
+}
+
+// translate input range information to input samples
+template <typename T>
+__global__ void ComputeInputSamples(const SampleDesc<T>* sample_data) {
+  const int64_t t_id = blockDim.x * blockDim.y * blockIdx.x + threadIdx.y * blockDim.x + threadIdx.x;
+  auto& sample = sample_data[blockIdx.y];
+  if (t_id >= sample.size_in) return;
+  sample.in[t_id] = sample.span.begin + (T)t_id / sample.span.sampling_rate;
+}
+
+template <typename T, template <typename> class W >
+DLL_PUBLIC KernelRequirements WaveletGpu<T, W>::Setup(KernelContext &context,
+                                                      const InListGPU<T> &a,
+                                                      const InListGPU<T> &b,
+                                                      const WaveletSpan<T> &span,
+                                                      const std::vector<T> &args) {
+  ENFORCE_SHAPES(a.shape, b.shape);
+  auto out_shape = this->GetOutputShape(a.shape, b.shape, span);
+  KernelRequirements req;
+  req.output_shapes = {out_shape};
+  wavelet_ = W(args);
+  return req;
+}
+
+template <typename T, template <typename> class W >
+DLL_PUBLIC void WaveletGpu<T, W>::Run(KernelContext &ctx,
+                                      OutListGPU<T> &out,
+                                      const InListGPU<T> &a,
+                                      const InListGPU<T> &b,
+                                      const WaveletSpan<T> &span) {
+  ENFORCE_SHAPES(a.shape, b.shape);
+
+  auto num_samples = a.num_samples();
+  std::vector<SampleDesc<T>> sample_data = std::vector<SampleDesc<T>>(num_samples);
+  int64_t max_size_in = 0, max_size_a = 0;
+
+  for (int i = 0; i < num_samples; i++) {
+    auto &sample = sample_data[i];
+    sample.out = out.tensor_data(i);
+    sample.a = a.tensor_data(i);
+    sample.size_a = a.shape.tensor_size(i);
+    max_size_a = std::max(max_size_a, sample.size_a);
+    sample.b = b.tensor_data(i);
+    sample.size_b = b.shape.tensor_size(i);
+    sample.span = span;
+    sample.size_in = std::ceil((sample.span.end - sample.span.begin) * sample.span.sampling_rate);
+    CUDA_CALL(cudaMalloc(&(sample.in), sizeof(T) * sample.size_in));
```

Comment:
Not needed I guess.

Reply:
I mean it either could be a host memory copied to the GPU later or scratchpad should be used for this allocation, not slow cudaMalloc.

Reply:
Changed cudaMalloc to scratchpad.AllocateGPU.